File: cronjobs-module/node/schedule.js
// schedule.js — deterministically add a job to data/jobs.json.
//
// Your twin calls THIS (via Bash, from the schedule-task skill) instead of
// hand-editing JSON, so scheduling is reliable. scheduler.js fires the job.
//
// Usage:
//   node schedule.js --at "2026-06-20T19:15:00" --message "hi!"
//   node schedule.js --at "2026-06-20T19:15:00" --prompt "Write me a hi message"
//   node schedule.js --cron "30 9 * * 1-5" --message "standup!"
//   optional: --title "label"  --channel "C0123 or U0123"
//
// --at is LOCAL time, ISO, no timezone suffix. --prompt lets the twin compose
// the message in your voice at fire time; --message is exact text.

const fs = require("fs");
const path = require("path");
const { randomUUID } = require("crypto");

const JOBS_FILE = process.env.JOBS_FILE
  ? path.resolve(process.env.JOBS_FILE)
  : path.join(__dirname, "data", "jobs.json");

function parseArgs(argv) {
  const flags = {};
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a.startsWith("--")) {
      const key = a.slice(2);
      const next = argv[i + 1];
      if (next === undefined || next.startsWith("--")) flags[key] = true;
      else {
        flags[key] = next;
        i++;
      }
    }
  }
  return flags;
}

const flags = parseArgs(process.argv.slice(2));

if (!flags.at && !flags.cron) {
  console.error('Need --at "<ISO local time>" or --cron "<expr>".');
  process.exit(1);
}
if (!flags.message && !flags.prompt) {
  console.error("Need --message <text> or --prompt <instruction>.");
  process.exit(1);
}

const job = {
  id: randomUUID().slice(0, 8),
  title: flags.title || (flags.message ? String(flags.message).slice(0, 40) : "scheduled task"),
  status: "pending",
  lastRun: null,
};
if (flags.at) job.runAt = String(flags.at);
if (flags.cron) job.cron = String(flags.cron);
if (flags.prompt) job.prompt = String(flags.prompt);
if (flags.message) job.message = String(flags.message);
if (flags.channel) job.channel = String(flags.channel);

fs.mkdirSync(path.dirname(JOBS_FILE), { recursive: true });
let jobs = [];
try {
  const raw = fs.readFileSync(JOBS_FILE, "utf8").trim();
  if (raw) jobs = JSON.parse(raw);
  if (!Array.isArray(jobs)) jobs = [];
} catch (_) {
  jobs = [];
}
jobs.push(job);
fs.writeFileSync(JOBS_FILE, JSON.stringify(jobs, null, 2));

console.log(
  `scheduled "${job.title}" ${job.runAt ? `at ${job.runAt}` : `on cron "${job.cron}"`} (id ${job.id})`,
);

File: cronjobs-module/python/schedule.py
"""schedule.py — deterministically add a job to data/jobs.json.

Your twin calls THIS (via Bash, from the schedule-task skill) instead of
hand-editing JSON, so scheduling is reliable. scheduler.py fires the job.

Usage:
  python schedule.py --at "2026-06-20T19:15:00" --message "hi!"
  python schedule.py --at "2026-06-20T19:15:00" --prompt "Write me a hi message"
  python schedule.py --cron "30 9 * * 1-5" --message "standup!"
  optional: --title "label"  --channel "C0123 or U0123"

--at is LOCAL time, ISO, no timezone suffix. --prompt lets the twin compose the
message in your voice at fire time; --message is exact text.
"""

import json
import os
import sys
import uuid
from pathlib import Path

JOBS_FILE = Path(os.environ.get("JOBS_FILE", Path(__file__).resolve().parent / "data" / "jobs.json"))


def parse_args(argv):
    flags = {}
    i = 0
    while i < len(argv):
        a = argv[i]
        if a.startswith("--"):
            key = a[2:]
            nxt = argv[i + 1] if i + 1 < len(argv) else None
            if nxt is None or nxt.startswith("--"):
                flags[key] = True
            else:
                flags[key] = nxt
                i += 1
        i += 1
    return flags


flags = parse_args(sys.argv[1:])

if not flags.get("at") and not flags.get("cron"):
    print('Need --at "<ISO local time>" or --cron "<expr>".', file=sys.stderr)
    sys.exit(1)
if not flags.get("message") and not flags.get("prompt"):
    print("Need --message <text> or --prompt <instruction>.", file=sys.stderr)
    sys.exit(1)

job = {
    "id": uuid.uuid4().hex[:8],
    "title": flags.get("title") or (str(flags["message"])[:40] if flags.get("message") else "scheduled task"),
    "status": "pending",
    "lastRun": None,
}
if flags.get("at"):
    job["runAt"] = str(flags["at"])
if flags.get("cron"):
    job["cron"] = str(flags["cron"])
if flags.get("prompt"):
    job["prompt"] = str(flags["prompt"])
if flags.get("message"):
    job["message"] = str(flags["message"])
if flags.get("channel"):
    job["channel"] = str(flags["channel"])

JOBS_FILE.parent.mkdir(parents=True, exist_ok=True)
jobs = []
try:
    raw = JOBS_FILE.read_text().strip()
    if raw:
        jobs = json.loads(raw)
    if not isinstance(jobs, list):
        jobs = []
except Exception:
    jobs = []
jobs.append(job)
JOBS_FILE.write_text(json.dumps(jobs, indent=2))

where = f'at {job["runAt"]}' if job.get("runAt") else f'on cron "{job["cron"]}"'
print(f'scheduled "{job["title"]}" {where} (id {job["id"]})')
